[resubmit][rpc] Add local shutdown to process group agent

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#30330 [resubmit][rpc] Add local shutdown to process group agent**
[test all]
Resubmit of https://github.com/pytorch/pytorch/pull/30020, which was reverted.
This is now possible due to previous changes made in `gloo` and `ProcessGroupGloo`. We `abort` the listener thread that is waiting for a message, and join all other threads. The destructor calls this same `localShutdown` method, but we ensure this is not called multiple times.


Differential Revision: [D18661775](https://our.internmc.facebook.com/intern/diff/D18661775/)

rohan-varma committed:
[resubmit][rpc] Add local shutdown to process group agent

This is now possible due to previous changes made in `gloo` and `ProcessGroupGloo`. We `abort` the listener thread that is waiting for a message, and join all other threads. The destructor calls this same `localShutdown` method, but we ensure this is not called multiple times.

ghstack-source-id: 94442415

Differential Revision: [D18661775](https://our.internmc.facebook.com/intern/diff/D18661775/)

[ghstack-poisoned]

mrshenli commented on test/rpc_test.py:
```diff
@@ -1090,6 +1104,50 @@ def test_call_method_on_rref(self):
+            rank=self.rank,
+            world_size=self.world_size,
+            init_method=self.init_method,
+            rpc_agent_options=self.rpc_agent_options,
```
IIUC, `self. rpc_agent_options` no longer exists, this needs to be replaced by  `rpc_backend_options`?

rohan-varma replied:
Sorry, looks like the update was coming just as you reviewed

rohan-varma committed:
Update on "[resubmit][rpc] Add local shutdown to process group agent"


This is now possible due to previous changes made in `gloo` and `ProcessGroupGloo`. We `abort` the listener thread that is waiting for a message, and join all other threads. The destructor calls this same `localShutdown` method, but we ensure this is not called multiple times.


Differential Revision: [D18661775](https://our.internmc.facebook.com/intern/diff/D18661775/)

[ghstack-poisoned]

zhaojuanmao commented:
@rohan-varma and @mrshenli I'm wondering why we need to split into wait_all_workers() and shutdown() in this way if shutdown() always needs to follow by wait_all_workers()?  looks like they are not independent.

mrshenli commented:
@zhaojuanmao 

One reason is that `wait_all_workers` could become a non-terminating API in the future, sth similar to `dist.barrier()`. So that applications can do some rpc/remote and then call `wait_all_workers` for a global sync, and then resume, and then maybe another global sync, and then `shutdown`. (as mentioned by @satgera) We don't explicitly say we support this yet, but we can discuss if that's what we want.

rohan-varma committed:
Update on "[resubmit][rpc] Add local shutdown to process group agent"


Resubmit of https://github.com/pytorch/pytorch/pull/30020, which was reverted.
This is now possible due to previous changes made in `gloo` and `ProcessGroupGloo`. We `abort` the listener thread that is waiting for a message, and join all other threads. The destructor calls this same `localShutdown` method, but we ensure this is not called multiple times.


Differential Revision: [D18661775](https://our.internmc.facebook.com/intern/diff/D18661775/)

[ghstack-poisoned]

rohan-varma commented on torch/csrc/distributed/rpc/process_group_agent.cpp:
```diff
@@ -269,14 +250,33 @@ void ProcessGroupAgent::sync() {
+      recvWork_->abort();
+    }
+  }
+  threadPool_.waitWorkComplete();
```
I had to add this here in shutdown, otherwise the test `test_rpc_shutdown()` would flake with gloo connection reset errors. I don't think it delays the shutdown by too much, as it just waits for already enqueued work to complete.

mrshenli replied:
hmm, does this need to be placed before abort? Otherwise, if there are indeed unfinished task, the receiving end listener thread might have already been aborted?

rohan-varma replied:
Hm, my reasoning for doing it after was that if it was before, we could possibly get more work while aborting the listenerThread, and then enqueue that recv, and not wait for it to be completed. For example, say that recvWork->wait has already been unblocked before we call its abort (via a message from another worker), we may have called enqueueRecv which will add a task to the thread pool.

zhaojuanmao commented:
lgtm!

xush6528 commented on torch/distributed/rpc/api.py:
```diff
@@ -51,7 +51,7 @@ def wrapper(*args, **kwargs):
     r"""
     Block until all local and remote RPC processes reach this method, and then
-    destroy local the RPC agent. Every RPC process must call this method before
+    destroy RRef and RPC handlers. Every RPC process must call this method before
```
"RPC handlers" -> "Python RPC handlers"


RPC handlers is vague.

rohan-varma replied:
This docstring is also in the wrong place now, will update

mrshenli replied:
> I actually don't think we need to mention that we destroy RPC handlers and RRef context, those are internal details that should be abstracted away from the user.

Agree, users don't need to know there is a RRef context and handler

rohan-varma committed:
Update on "[resubmit][rpc] Add local shutdown to process group agent"


Resubmit of https://github.com/pytorch/pytorch/pull/30020, which was reverted.
This is now possible due to previous changes made in `gloo` and `ProcessGroupGloo`. We `abort` the listener thread that is waiting for a message, and join all other threads. The destructor calls this same `localShutdown` method, but we ensure this is not called multiple times.


Differential Revision: [D18661775](https://our.internmc.facebook.com/intern/diff/D18661775/)

[ghstack-poisoned]

mrshenli commented on torch/distributed/rpc/api.py:
```diff
@@ -64,24 +65,38 @@ def wait_all_workers():
         >>> rpc.wait_all_workers()
+        >>> rpc.shutdown()
     """
     global _agent
```
(This can come in a followup PR)

We might not need this `global _agent` any more as we are not modifying it.

mrshenli reviewed (approved):
Approve to unblock. Please wait for all tests to pass.

rohan-varma committed:
Update on "[test all][resubmit][rpc] Add local shutdown to process group agent"

[test all]
Resubmit of https://github.com/pytorch/pytorch/pull/30020, which was reverted.
This is now possible due to previous changes made in `gloo` and `ProcessGroupGloo`. We `abort` the listener thread that is waiting for a message, and join all other threads. The destructor calls this same `localShutdown` method, but we ensure this is not called multiple times.


Differential Revision: [D18661775](https://our.internmc.facebook.com/intern/diff/D18661775/)

[ghstack-poisoned]

xush6528 commented on torch/csrc/distributed/rpc/process_group_agent.cpp:
```diff
@@ -273,14 +254,33 @@ void ProcessGroupAgent::sync() {
+  }
+  threadPool_.waitWorkComplete();
+  listenerThread_.join();
+  futureTimeoutCV_.notify_one();
```
What if the `futureTimeoutThread_` is not blocked on the `futureTimeoutCV_`? For example, it could be at line `std::chrono::milliseconds sleepTime;`.

Will `futureTimeoutThread_` live forever?

rohan-varma replied:
Thanks for pointing this out - so it sounds like this should be fixed if we hold the lock on `rpcRunning`? Should we hold that lock wherever we write/read it?

xush6528 commented on torch/csrc/distributed/rpc/process_group_agent.cpp:
```diff
@@ -273,14 +254,33 @@ void ProcessGroupAgent::sync() {
 }
 
 void ProcessGroupAgent::start() {
+  rpcRunning_.store(true);
```
I changed the `futureTimeoutCV_.wait_for(...)` part in https://github.com/pytorch/pytorch/pull/30355.

We need to change this line to
```
std::lock_guar<std::mutex> futureLock{futureMutex_};
rpcRunning_.store(true);
```
in #30355 as well.

xush6528 commented on torch/csrc/distributed/rpc/process_group_agent.cpp:
```diff
@@ -273,14 +254,33 @@ void ProcessGroupAgent::sync() {
+void ProcessGroupAgent::shutdown() {
+  if (!rpcRunning_.exchange(false)) {
+    return;
+  }
```
Here as well.

```cpp
std::lock_guar<std::mutex> futureLock{futureMutex_};
if (!rpcRunning_.exchange(false)) {
  return;
}
```

rohan-varma replied:
Added.

mrshenli reviewed (changes requested):
Remove stamp until we address @xush6528's comment

rohan-varma committed:
Update on "[resubmit][rpc] Add local shutdown to process group agent"

[test all]
Resubmit of https://github.com/pytorch/pytorch/pull/30020, which was reverted.
This is now possible due to previous changes made in `gloo` and `ProcessGroupGloo`. We `abort` the listener thread that is waiting for a message, and join all other threads. The destructor calls this same `localShutdown` method, but we ensure this is not called multiple times.


Differential Revision: [D18661775](https://our.internmc.facebook.com/intern/diff/D18661775/)

[ghstack-poisoned]

rohan-varma commented on torch/csrc/distributed/rpc/process_group_agent.cpp:
```diff
@@ -487,9 +504,12 @@ void ProcessGroupAgent::listenLoop() {
-    std::chrono::milliseconds sleepTime;
+  while (true) {
     std::unique_lock<std::mutex> lock{futureMutex_};
+    if (!rpcRunning_.load()) {
```
We now hold the lock from when we check `rpcRunning` to until we sleep, and the lock is also held in `shutdown()` before we set `rpcRunning`.

mrshenli commented on torch/csrc/distributed/rpc/process_group_agent.cpp:
```diff
@@ -487,9 +504,12 @@ void ProcessGroupAgent::listenLoop() {
-  while (!shutdown_.load()) {
-    std::chrono::milliseconds sleepTime;
+  while (true) {
     std::unique_lock<std::mutex> lock{futureMutex_};
```
Just make sure I understand this correctly. If this is before the setting `rpcRunning_` in `shutdown`, then `futureTimeoutCV_` will be notified properly. If this is after, then this would return as `rpcRunning_.load() == false`, right?

rohan-varma replied:
Yes - for the first case you mentioned, this thread would drop the lock when it waits and would be appropriately notified by shutdown. For the 2nd case, we would break as `rpcRunning = false`.

mrshenli reviewed (approved):
Thanks for fixing promptly!

rohan-varma committed:
Update on "[resubmit][rpc] Add local shutdown to process group agent"

[test all]
Resubmit of https://github.com/pytorch/pytorch/pull/30020, which was reverted.
This is now possible due to previous changes made in `gloo` and `ProcessGroupGloo`. We `abort` the listener thread that is waiting for a message, and join all other threads. The destructor calls this same `localShutdown` method, but we ensure this is not called multiple times.


Differential Revision: [D18661775](https://our.internmc.facebook.com/intern/diff/D18661775/)

[ghstack-poisoned]

rohan-varma commented on torch/csrc/distributed/rpc/process_group_agent.cpp:
```diff
@@ -509,7 +529,7 @@ void ProcessGroupAgent::pollTimedOutRPCs() {
     }
 
-    if (shutdown_.load()) {
+    if (!rpcRunning_.load()) {
```
This check is not strictly necessary, but can help performance. If the `shutdown` thread wakes up the CV here, then `rpcRunning` will always be false, without this check, we would continue to `processTimedOutFutures`, which would be a no-op.

pietern commented on test/rpc_test.py:
```diff
@@ -18,6 +19,20 @@
+
+def _check_rpc_done(rank_distance):
+    while not rpc_done[rank_distance]:
+        time.sleep(0.1)
```
This can do `time.sleep(0)` to yield control and effectively busy wait.

pietern commented on torch/csrc/distributed/rpc/process_group_agent.h:
```diff
@@ -161,12 +159,19 @@ class ProcessGroupAgent : public RpcAgent {
+  // ProcessGroupAgent::start and unset in ProcessGroupAgent::shutdown and
+  // ProcessGroupAgent::join. It controls whether several background threads
+  // should be running.
+  std::atomic<bool> rpcRunning_{false};
```
Since you're locking access anyway and not using this as an atomic, I recommend changing this to a regular bool to avoid the impression you can use it as atomic.

rohan-varma replied:
Synced with @xush6528 about this - currently, it is not always locked (see the checks in `listenLoop` and `send`). We only need to strictly lock it when it is used in conjunction with the `futureTimeout` condvar, so it might be worth leaving it as atomic?

pietern replied:
Sounds good, but please document what the expected locking strategy is.

Using an atomic implies (at least to me) that you don't need locks. To reduce the probability of this causing breakage later on, I think a regular bool is better. Accessing that without locks is a red flag in code review.

rohan-varma replied:
Will add this in a follow up PR.

rohan-varma replied:
Ended up doing it in this PR since had to do a rebase anyways.

xush6528 commented on torch/csrc/distributed/rpc/process_group_agent.cpp:
```diff
@@ -487,9 +504,12 @@ void ProcessGroupAgent::listenLoop() {
+    if (!rpcRunning_.load()) {
+      return;
+    }
+    std::chrono::milliseconds sleepTime;
```
I am not convinced yet. How do you make sure the notification is always delivered with no cond var predicate added?

e.g. What if `pollTimedOutRPCs` is at here, and `shutdown` sends a notification? Would that notification be lost?

rohan-varma replied:
For `shutdown` to have sent a notification, this implies that it would have had to hold the `futureMutex` lock and set `rpcRunning` to be false before. But since that now requires the lock, we can't have `shutdown` be at `std::chrono::milliseconds sleepTime;` at the same time, since this section is also locked.

jjlilley commented on torch/csrc/distributed/rpc/process_group_agent.cpp:
```diff
@@ -460,10 +465,22 @@ void ProcessGroupAgent::enqueueRecv(RecvWork work) {
-    pg_->recvAnysource(preamble, pg_->getRank())->wait();
+    {
+      std::lock_guard<std::mutex> guard(recvWorkMutex_);
+      recvWork_ = pg_->recvAnysource(preamble, pg_->getRank());
```
why do you need to hold the mutex across pg_->recvAnySource()?

rohan-varma replied:
We want to make sure that `recvWork` is in a valid initialized state when we call `abort` on it from the shutdown thread, so we'd want the lock to ensure there is no race when it is initialized here and when it is checked in the `shutdown` function.

jjlilley replied:
Ah, I see - I initially didn't see that this was a separate mutex. Thanks.

fwiw, it's a little sketchy to access recvWork_ on 479 without holding the mutex, you should ideally keep a local ref to the shared_ptr, e.g.

auto work = pg_->recvAnysource(preamble, pg_->getRank());
setPendingRecvWork(work);  // or put inline, just grabs lock and sets recvWork_
if (!rpcRunning_.load() ||
   !work->wait() /* aborted */) {
   return;
}

If nothing else, it's easier to reason about shutdown safety, when we guarantee that recvWork_ is only accessed under the lock.

rohan-varma replied:
Thanks for calling this out, this is implemented in the latest version.

jjlilley replied:
Great!
fwiw, LGTM

rohan-varma committed:
Update on "[resubmit][rpc] Add local shutdown to process group agent"

[test all]
Resubmit of https://github.com/pytorch/pytorch/pull/30020, which was reverted.
This is now possible due to previous changes made in `gloo` and `ProcessGroupGloo`. We `abort` the listener thread that is waiting for a message, and join all other threads. The destructor calls this same `localShutdown` method, but we ensure this is not called multiple times.


Differential Revision: [D18661775](https://our.internmc.facebook.com/intern/diff/D18661775/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[resubmit][rpc] Add local shutdown to process group agent"

[test all]
Resubmit of https://github.com/pytorch/pytorch/pull/30020, which was reverted.
This is now possible due to previous changes made in `gloo` and `ProcessGroupGloo`. We `abort` the listener thread that is waiting for a message, and join all other threads. The destructor calls this same `localShutdown` method, but we ensure this is not called multiple times.


Differential Revision: [D18661775](https://our.internmc.facebook.com/intern/diff/D18661775/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[resubmit][rpc] Add local shutdown to process group agent"

[test all]
Resubmit of https://github.com/pytorch/pytorch/pull/30020, which was reverted.
This is now possible due to previous changes made in `gloo` and `ProcessGroupGloo`. We `abort` the listener thread that is waiting for a message, and join all other threads. The destructor calls this same `localShutdown` method, but we ensure this is not called multiple times.


Differential Revision: [D18661775](https://our.internmc.facebook.com/intern/diff/D18661775/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[resubmit][rpc] Add local shutdown to process group agent"

[test all]
Resubmit of https://github.com/pytorch/pytorch/pull/30020, which was reverted.
This is now possible due to previous changes made in `gloo` and `ProcessGroupGloo`. We `abort` the listener thread that is waiting for a message, and join all other threads. The destructor calls this same `localShutdown` method, but we ensure this is not called multiple times.


Differential Revision: [D18661775](https://our.internmc.facebook.com/intern/diff/D18661775/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[resubmit][rpc] Add local shutdown to process group agent"

[test all]
Resubmit of https://github.com/pytorch/pytorch/pull/30020, which was reverted.
This is now possible due to previous changes made in `gloo` and `ProcessGroupGloo`. We `abort` the listener thread that is waiting for a message, and join all other threads. The destructor calls this same `localShutdown` method, but we ensure this is not called multiple times.


Differential Revision: [D18661775](https://our.internmc.facebook.com/intern/diff/D18661775/)

[ghstack-poisoned]

mrshenli commented on torch/distributed/rpc/api.py:
```diff
@@ -63,27 +84,28 @@ def wait_all_workers():
+        if graceful:
+            _wait_all_workers()
+        _agent.shutdown()
         _destroy_rref_context()
```
When you rebase onto #30217, could you please please move this line to the front of `_agent.shutdown()`. This shouldn't make any difference in our tests for now. However, `ctx.destroyed_` is only set to `true` in `_destroy_rref_context()`, meaning that logically, RRef destructions could still trigger messages been sent using the `_agent` before `_destroy_rref_context()`.

rohan-varma replied:
Should it be before `_wait_all_workers` as well then, since we want to wait for these messages to be processed?

mrshenli replied:
Good point! Yes, it should!

rohan-varma committed:
Update on "[resubmit][rpc] Add local shutdown to process group agent"

[test all]
Resubmit of https://github.com/pytorch/pytorch/pull/30020, which was reverted.
This is now possible due to previous changes made in `gloo` and `ProcessGroupGloo`. We `abort` the listener thread that is waiting for a message, and join all other threads. The destructor calls this same `localShutdown` method, but we ensure this is not called multiple times.


Differential Revision: [D18661775](https://our.internmc.facebook.com/intern/diff/D18661775/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[resubmit][rpc] Add local shutdown to process group agent"

[test all]
Resubmit of https://github.com/pytorch/pytorch/pull/30020, which was reverted.
This is now possible due to previous changes made in `gloo` and `ProcessGroupGloo`. We `abort` the listener thread that is waiting for a message, and join all other threads. The destructor calls this same `localShutdown` method, but we ensure this is not called multiple times.


Differential Revision: [D18661775](https://our.internmc.facebook.com/intern/diff/D18661775/)

[ghstack-poisoned]

rohan-varma commented:
cc @pritamdamania87, this is the doc update: 
<img width="888" alt="doc_update" src="https://user-images.githubusercontent.com/8039770/69684932-629b9880-106f-11ea-8629-f5c664c4a214.png">

rohan-varma committed:
Update on "[resubmit][rpc] Add local shutdown to process group agent"

[test all]
Resubmit of https://github.com/pytorch/pytorch/pull/30020, which was reverted.
This is now possible due to previous changes made in `gloo` and `ProcessGroupGloo`. We `abort` the listener thread that is waiting for a message, and join all other threads. The destructor calls this same `localShutdown` method, but we ensure this is not called multiple times.


Differential Revision: [D18661775](https://our.internmc.facebook.com/intern/diff/D18661775/)

[ghstack-poisoned]

rohan-varma commented on torch/distributed/rpc/api.py:
```diff
@@ -73,27 +95,28 @@ def wait_all_workers():
         # cleanup() function has python dependency, it assumes python
         # interpreter exists
         _cleanup_python_rpc_handler()
+        _agent = None
```
cc @mrshenli this is the new shutdown sequence after the ignore rref leaks PR has landed

mrshenli replied:
this looks good to me, thanks!

rohan-varma committed:
Update on "[resubmit][rpc] Add local shutdown to process group agent"

[test all]
Resubmit of https://github.com/pytorch/pytorch/pull/30020, which was reverted.
This is now possible due to previous changes made in `gloo` and `ProcessGroupGloo`. We `abort` the listener thread that is waiting for a message, and join all other threads. The destructor calls this same `localShutdown` method, but we ensure this is not called multiple times.


Differential Revision: [D18661775](https://our.internmc.facebook.com/intern/diff/D18661775/)

[ghstack-poisoned]

pietern commented on docs/source/notes/distributed_autograd.rst:
```diff
@@ -330,50 +331,52 @@ autograd and distributed optimizer is as follows:
+  # Run world_size workers.
+  world_size = 2
+  for i in range(world_size):
+      p = mp.Process(target=run_process, args=(i, (i + 1) % 2, world_size))
```
`mp.spawn` for the win!

rohan-varma replied:
Addressing in separate PR (https://github.com/pytorch/pytorch/pull/30381)

pietern commented on test/rpc_test.py:
```diff
@@ -18,6 +19,20 @@
+
+def _check_rpc_done(rank_distance):
+    while not rpc_done[rank_distance]:
+        time.sleep(0)
```
Could add a comment here to say that this yields control to other threads so it won't be removed.

rohan-varma replied:
Will address in a follow up PR.

rohan-varma replied:
I ended up addressing it in this PR, since I had to do a rebase anyways.

rohan-varma commented:
The following are the CI errors, all unrelated:

1) Error response from daemon: manifest for 308535385114.dkr.ecr.us-east-1.amazonaws.com/pytorch/pytorch-linux-xenial-py3-clang5-android-ndk-r19c:405-3796e2c20d4587ffdcae63b42490a63316be44bd-android-arm-v7a not found

2) Nov 27 08:55:40 FAIL [0.032s]: test_neg_cuda (__main__.TestTorchDeviceTypeCUDA)
Nov 27 08:55:40 ----------------------------------------------------------------------
Nov 27 08:55:40 RuntimeError: CUDA error: misaligned address

3) Nov 27 00:01:04 Traceback (most recent call last):
Nov 27 00:01:04   File "/Users/distiller/workspace/miniconda3/lib/python3.7/site-packages/soundfile.py", line 142, in <module>
Nov 27 00:01:04     raise OSError('sndfile library not found')
Nov 27 00:01:04 OSError: sndfile library not found

4) 08:34:39 FAIL: test_logical_or_cuda (__main__.TestTorchDeviceTypeCUDA)
08:34:39 ----------------------------------------------------------------------
08:34:39 Traceback (most recent call last):
08:34:39   File "/var/lib/jenkins/workspace/test/common_utils.py", line 631, in wrapper
08:34:39     method(*args, **kwargs)
08:34:39   File "/var/lib/jenkins/workspace/test/common_device_type.py", line 179, in instantiated_test
08:34:39     return test(self, device_arg)
08:34:39   File "test_torch.py", line 6578, in test_logical_or
08:34:39     self._test_logical(device, 'logical_or', [10, 0, 1, 0], [1, 0, 0, 10], [1, 0, 1, 1])
08:34:39   File "test_torch.py", line 6555, in _test_logical
08:34:39     self.assertEqual(expected_res.bool(), getattr(a, op)(b))
08:34:39   File "/var/lib/jenkins/workspace/test/common_utils.py", line 808, in assertEqual
08:34:39     assertTensorsEqual(x, y)
08:34:39   File "/var/lib/jenkins/workspace/test/common_utils.py", line 778, in assertTensorsEqual
08:34:39     self.assertLessEqual(max_err, prec, message)
08:34:39 AssertionError: tensor(1, device='cuda:0', dtype=torch.int32) not less than or equal to 1e-05 :

rohan-varma committed:
Update on "[resubmit][rpc] Add local shutdown to process group agent"

[test all]
Resubmit of https://github.com/pytorch/pytorch/pull/30020, which was reverted.
This is now possible due to previous changes made in `gloo` and `ProcessGroupGloo`. We `abort` the listener thread that is waiting for a message, and join all other threads. The destructor calls this same `localShutdown` method, but we ensure this is not called multiple times.


Differential Revision: [D18661775](https://our.internmc.facebook.com/intern/diff/D18661775/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[resubmit][rpc] Add local shutdown to process group agent"

[test all]
Resubmit of https://github.com/pytorch/pytorch/pull/30020, which was reverted.
This is now possible due to previous changes made in `gloo` and `ProcessGroupGloo`. We `abort` the listener thread that is waiting for a message, and join all other threads. The destructor calls this same `localShutdown` method, but we ensure this is not called multiple times.


Differential Revision: [D18661775](https://our.internmc.facebook.com/intern/diff/D18661775/)

[ghstack-poisoned]

rohan-varma commented:
CI failure that is not listed as "broken upstream" has occured on other merged PRs (https://app.circleci.com/jobs/github/pytorch/pytorch/3733973,  https://app.circleci.com/jobs/github/pytorch/pytorch/3734258), it should be fixed by this revert: https://github.com/pytorch/pytorch/commit/a2ed50c9209cb61eb72465b55936c4bdc468cc35
